[Benchmark] Add sampling parameters to benchmark_serving.

hyeygit opened the pull request:
Allow specifying sampling params (top-k, top-p etc) in the online benchmark. This is done by adding the sampling params to the "extra_body" field in the client request.

New command line flags (`--top-p`, `--top-k`, `--min-p`, and `--temperature`) are added to `benchmark_serving.py`.

### Example benchmark results

After locally enabling top-k and top-p sampling for TPU, I ran the serving benchmark on v6e-1, Llama3.1-8B, with `--top-k=10` and `--top-p=0.8`. Here are the results before/after the TPU optimization. The no-sampling baseline is also included for comparison.

|Description                                        |Throughput (req/s)|
|---------------------------------------------------|------------------|
|Baseline (NO sampling)                             |5.98              |
|Baseline w/ non-optimized sampling (forward_native)|0.24              |
|TPU-optimized sampling                             |5.57              |

hyeygit committed:
Add sampling parameters to benchmark_serving.

Allow specifying sampling params (top-k, top-p etc) in the online benchmark.
This is done by adding the params to the "extra_body" field in the client
request.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

simon-mo commented:
Can you explain why is this needed? sampling algorithms shouldn't change the perf result much.

hyeygit commented:
> Can you explain why is this needed? sampling algorithms shouldn't change the perf result much.

@simon-mo Sampling (specifically top-k and top-p) can be slow on certain backend such as XLA/TPU. We recently optimized top-p sampling for TPU to make it orders of magnitude faster (#15736). It will be nice to see the impact on end-to-end serving performance (and based on my local testing the top-p optimization sped up serving throughput by 23X).

ywang96 commented:
@JenZhao Can you give this PR a first pass? Thanks!

JenZhao commented:
Thank you! It looks good to me. Could you provide some examples that demonstrate the impact of sampling on performance results, and also update the README to include this change?

ywang96 commented on benchmarks/benchmark_serving.py:
```diff
@@ -1000,6 +1018,24 @@ def main(args: argparse.Namespace):
     )
 
+    sampling_group = parser.add_argument_group("sampling parameters")
+    sampling_group.add_argument("--top-p",
+                                type=float,
+                                default=None,
+                                help="Top-p sampling parameter.")
+    sampling_group.add_argument("--top-k",
+                                type=int,
+                                default=None,
+                                help="Top-k sampling parameter.")
+    sampling_group.add_argument("--min-p",
+                                type=float,
+                                default=None,
+                                help="Min-p sampling parameter.")
+    sampling_group.add_argument("--temperature",
+                                type=float,
+                                default=None,
+                                help="Temperature sampling parameter.")
```
These are only passed to `async_request_openai_completions` and `async_request_openai_chat_completions`. Please update the help messages here to reflect this and IMO we should add an assertion to make sure users are only using openai completions or chat completions whenever any of these is specified.

hyeygit replied:
Thanks for the review! Good catch! Done.

ywang96 commented on benchmarks/benchmark_serving.py:
```diff
@@ -1000,6 +1018,24 @@ def main(args: argparse.Namespace):
+                                type=float,
+                                default=None,
+                                help="Min-p sampling parameter.")
+    sampling_group.add_argument("--temperature",
+                                type=float,
+                                default=None,
+                                help="Temperature sampling parameter.")
```
I think it's probably better if we set a default value of `0.0` here for temperature, or otherwise indicate in the help message that greedy decoding is used if temperature is not specified`. WDYT?

hyeygit replied:
SG. Set the default to 0.0 and also explained in help message.

ywang96 reviewed (commented):
@hyeygit I left two comments - PTAL

NickLucche commented on benchmarks/benchmark_serving.py:
```diff
@@ -652,6 +656,19 @@ def main(args: argparse.Namespace):
+            'top_p': args.top_p,
+            'top_k': args.top_k,
+            'min_p': args.min_p,
+            'temperature': args.temperature
```
Shouldn't we remove the default temperature value that appears in `async_request_openai_completions` and `async_request_openai_chat_completions`?
Since it gets replaced anyway every time, I feel it hurts readability a bit.

hyeygit replied:
Thanks for the review @NickLucche ! I think keeping the default temperature in `backend_request_func.py` might be better because this file is likely used by other modules as well and not just by `benchmark_serving.py`, so we probably shouldn't modify its default behavior.

NickLucche reviewed (changes requested):
Nice one @hyeygit , I think this is generally useful to test the e2e impact of different sampling implementations!

hyeygit committed:
Address review feedback

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

hyeygit committed:
Add sampling example to README

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

hyeygit committed:
Merge branch 'main' into bench_serving_no_base

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

hyeygit commented:
Thank you all for the review! Comments addressed. PTAL.

@JenZhao I updated the PR description with example benchmark result. Also updated the README.

ywang96 reviewed (approved):
LGTM! Thanks for the contribution
